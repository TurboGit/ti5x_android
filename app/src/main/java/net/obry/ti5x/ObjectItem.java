package net.obry.ti5x;

// another class to handle item's id and name

public class ObjectItem {

  private int itemId;
  private String itemName;

  // constructor
  public ObjectItem(int itemId, String itemName) {
    this.itemId = itemId;
    this.itemName = itemName;
  }
}
