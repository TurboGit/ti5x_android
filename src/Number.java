package net.obry.ti5x;

public class Number
{
    private double v = 0;
    private boolean error;

    public static final int ANG_RAD  = 1;
    public static final int ANG_DEG  = 2;
    public static final int ANG_GRAD = 3;
    public static final Number PI = new Number(Math.PI);
    public static final Number ONE = new Number(1.0);
    public static final Number TEN = new Number(10.0);
    public static final Number ZERO  = new Number(0.0);

    private final static double D_ZERO      = 0.0;
    private final static double HALF_PI   = Math.PI / 2.0;
    private final static double HALF3_PI  = 3.0 * Math.PI / 2.0;
    private final static double TWO_PI    = Math.PI * 2.0;
    private final static double FROM_DEG  = Math.PI / 180.0;
    private final static double FROM_RAD  = Math.PI / 200.0;
    private final static double ERROR     = 9.9999999e99;
    private final static double LOG10     = Math.log(10);
    private final static double LOG1000   = Math.log(1000);

    private final static Number N_PI       = new Number(Math.PI);
    private final static Number N_HALF_PI  = new Number(HALF_PI);
    private final static Number N_HALF3_PI = new Number(HALF3_PI);
    private final static Number N_TWO_PI   = new Number(TWO_PI);

    public Number()
    {
        v = 0.0;
        error = false;
    }

    public Number(double d)
    {
        v = d;
        error = false;
    }

    public Number(String s)
    {
        v =  Double.parseDouble (s);
        error = false;
    }

    public Number(Number n)
    {
        this.v = n.v;
        this.error = false;
    }

    public boolean isEqual(Number d)
    {
        final double Epsilon = 1e-11;

        if (Math.abs (v - d.v) <= Epsilon)
            return true;
        else
            return false;
    }

    public int compareTo(Number n)
    {
        if (v < n.v)
            return -1;
        else if (v > n.v)
            return 1;
        else
            return 0;
    }

    public int compareTo(int i)
    {
        if (v < (double)i)
            return -1;
        else if (v > (double)i)
            return 1;
        else
            return 0;
    }

    public boolean isError()
    {
        final boolean R = error;
        error = false;
        return R;
    }

    public boolean isNaN()
    {
        return Double.isNaN(v);
    }

    public boolean isInfinite()
    {
        return Double.isInfinite(v);
    }

    public void set(double d)
    {
        v = d;
        error = false;
    }

    public void set(Number n)
    {
        v = n.v;
        error = false;
    }

    public void set(String s)
    {
        v = Double.parseDouble (s);
        error = false;
    }

    public double get()
    {
        return v;
    }

    public long getInt()
    {
        final double Vabs = Math.abs(v);
        final double Epsilon =
            Vabs > 0.8
            ? 1.0 / Math.pow (10, 15 - 3 - Math.min(1, Math.abs((int)Math.floor(Math.log(Vabs) / Math.log(10.0)))))
            : 0.0;

        return (long)Math.floor(Vabs + Epsilon);
    }

    public int getSignum()
    {
        return (int)Math.signum(v);
    }

    public void signum()
    {
        v = Math.signum(v);
    }

    public void abs()
    {
        v = Math.abs(v);
    }

    public void negate()
    {
        v = -v;
    }

    public void min(long i)
    {
        double N = (double)i;
        if (v > N)
            v = N;
    }

    public void max(long i)
    {
        double N = (double)i;
        if (v < N)
            v = N;
    }

    public void add(long i)
    {
        v += (double)i;
    }

    public void add(Number n)
    {
        v += n.v;
    }

    public void sub(Number n)
    {
        v -= n.v;
    }

    public void sub(long i)
    {
        v -= (double)i;
    }

    public void mult(Number n)
    {
        v *= n.v;
    }

    public void mult(long i)
    {
        v *= (double)i;
    }

    public void div(Number n)
    {
        if (n.v == 0)
        {
            v = ERROR;
            error = true;
        }
        else
            v /= n.v;
    }

    public void div(long i)
    {
        div(new Number((double)i));
    }

    public void rem(Number n)
    {
        v = Math.IEEEremainder(v, n.v);
    }

    public void rem(long i)
    {
        v = Math.IEEEremainder(v, (double)i);
    }

    public void trigScale(int Angle)
    {
        v = trigScale2(Angle);
    }

    private static double trigScale2(int Angle)
    {
        switch (Angle)
        {
           case ANG_DEG:
               return FROM_DEG;
           case ANG_GRAD:
               return FROM_RAD;
           default:
               return 1.0;
        }
    }

    private void NormalizeAngle(int Angle)
    {
        v = v * trigScale2(Angle);

        if (v < 0)
        {
            while (v < 0.0)
                v += TWO_PI;
        }
        else if (v >= TWO_PI)
        {
            while (v >= TWO_PI)
                v -= TWO_PI;
        }
    }

    public void cos(int Angle)
    {
        NormalizeAngle(Angle);

        v = Math.cos(v);
        if (isEqual(ZERO))
            v = 0;
    }
    public void acos(int Angle)
    {
        v = Math.acos(v) / trigScale2(Angle);
    }

    public void sin(int Angle)
    {
        NormalizeAngle(Angle);

        v = Math.sin(v);
        if (isEqual(ZERO))
            v = D_ZERO;
    }
    public void asin(int Angle)
    {
        v = Math.asin(v) / trigScale2(Angle);
    }

    public void tan(int Angle)
    {
        NormalizeAngle(Angle);

        if (isEqual(N_HALF_PI) || isEqual(N_HALF3_PI))
        {
            v = ERROR;
            error = true;
        }
        else if (isEqual(ZERO) || isEqual(N_PI))
        {
            v = D_ZERO;
        }
        else
            v = Math.tan(v);
    }
    public void atan(int Angle)
    {
        v = Math.atan(v) / trigScale2(Angle);
    }

    public void log()
    {
        if (v < 0.0)
        {
            v = Math.log10(-v);
            error = true;
        }
        else if (v == 0.0)
        {
            v = -ERROR;
            error = true;
        }
        else
            v = Math.log10(v);
    }

    public void ln()
    {
        if (v < 0.0)
        {
            v = Math.log(-v);
            error = true;
        }
        else if (v == 0.0)
        {
            v = -ERROR;
            error = true;
        }
        else
            v = Math.log(v);
    }

    public void sqrt()
    {
        if (v < 0.0)
        {
            v = Math.sqrt(-v);
            error = true;
        }
        else
            v = Math.sqrt(v);
    }

    public void x2()
    {
        v *= v;
    }

    public void reciprocal()
    {
        if (v == 0)
        {
            v = ERROR;
            error = true;
        }
        else
            v = 1.0 / v;
    }

    public void pow(Number n)
    {
        v = Math.pow(v, n.v);
    }

    public void exp()
    {
        v = Math.exp(v);
    }

    public void intPart()
    {
        final double Vabs = Math.abs(v);
        final double Epsilon =
            Vabs > 0.8
            ? 1.0 / Math.pow (10, 15 - 3 - Math.min(1, Math.abs((int)Math.floor(Math.log(Vabs) / Math.log(10.0)))))
            : 0.0;
        v = Math.floor(Vabs + Epsilon);
    }

    public void fracPart()
    {
        final double i = (long)v;
        v = v - i;
    }

    public void Pi()
    {
        v = Math.PI;
    }

    /* returns the number of figures before the decimal point in the
       formatted representation of X scaled by Exp. This has to be
       at least 1, because the decimal point is part of the display
       of the preceding digit. */

    public int figuresBeforeDecimal(int Exp)
    {
        int BeforeDecimal;

        if (v != 0.0)
        {
            BeforeDecimal = Math.max
                ((int)Math.ceil(Math.log10(Math.abs(v) / Math.pow(10.0, Exp))), 1);
        }
        else
        {
            BeforeDecimal = 1;
        }

        return BeforeDecimal;
    }

    /* returns the exponent scale to display X using the given format. */

    public int scaleExp(int UsingFormat)
    {
        int Exp = 0;
        double l = Math.abs(v);
        if (l != 0.0)
        {
            switch (UsingFormat)
            {
            case State.FORMAT_FLOAT:
                Exp = (int)Math.floor(Math.log(l) / LOG10);
                break;
            case State.FORMAT_ENG:
                Exp = (int)Math.floor(Math.log(l) / LOG1000) * 3;
                break;
            }
        }
        return Exp;
    }

    public void nDigits(int n)
    {
        if (n >= 0)
        {
            final double Factor = Math.pow(10, n);
            v = (int)(v * Factor) / Factor;
        }
    }

    public String formatString(java.util.Locale locale, int nrDecimals)
    {
        return String.format(locale, String.format(locale, "%%.%df", nrDecimals), v);
    }
}
