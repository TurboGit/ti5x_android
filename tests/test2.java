
// Test random from original TI

import net.obry.ti5x.Number;
import net.obry.ti5x.State;

public class test2
{

    public static void main(String[] args)
    {
        Number n1, n2, n3;
        Number r9;
        Number x;
        Number e5;
        Number Pi = new Number();
        Pi.Pi();
        Number n6 = new Number(5);

        // test 1 - RANDOM

        n1 = new Number(24298);
        n2 = new Number(99991);
        n3 = new Number(199017);
        r9 = new Number();
        x = new Number();
        e5 = new Number();

        e5.set(10);
        e5.pow(n6);
        r9.set(Pi);

        for (int k=1; k<3; k++)
        {
            x.set(n1);
            x.mult(r9);
            x.add(n2);

            r9.set(n3);
            x.div(n3);

            x.fracPart();
            r9.mult(x);
            x.mult(e5);
            x.intPart();
            x.div(e5);

            System.out.println("test - " + k + " " + x.get());
        }

        Number orig = new Number(10);
        Number copy = new Number(orig);
        copy.x2();
        System.out.println("orig=" + orig.get());
        System.out.println("copy=" + copy.get());


        // test 2 - TRIG

        x.set(90);
        x.cos(Number.ANG_DEG);
        System.out.println("cos(90)=" + x.get());
        x.set(45);
        x.cos(Number.ANG_DEG);
        System.out.println("cos(45)=" + x.get());
        x.set(Number.PI);
        x.cos(Number.ANG_RAD);
        System.out.println("cos(pi)=" + x.get());

        x.set(8);
        x.sqrt();
        System.out.println("sqrt(8)=" + x.get() + " " + x.isError());
        x.set(0);
        x.sqrt();
        System.out.println("sqrt(0)=" + x.get() + " " + x.isError());
        x.set(-5);
        x.sqrt();
        System.out.println("sqrt(-5)=" + x.get() + " " + x.isError());

        // test 3 - POW/LOG

        x.set(98765);
        x.log();
        System.out.println("log(98765)=" + x.get() + " " + x.isError());
        x.set(98765);
        x.ln();
        System.out.println("ln(98765)=" + x.get() + " " + x.isError());

        x.set(10);
        n6.set(2.5);
        x.pow(n6);
        System.out.println("pow(2.5)=" + x.get() + " " + x.isError());  // ?? wrong value should be 316.2277660141

        // test 4 - PRECISION

        Number one = new Number(1.0);
        Number lc = new Number(0.05);
        Number n4 = new Number(0.1);
        Number n5 = new Number(1);

        for (int k=0; k<100; k++)
            {
                one.add(lc);
                one.add(n4);
                one.sub(lc);
            }

        n4.set(0.1 * 100);
        one.sub(n4);

        if (one.isEqual(n5))
            System.out.println("one = " + one.get());
        else
            System.out.println("not one = " + one.get());

        // test 5 - figures

        x.set(123.456);
        System.out.println("fbd (123.456) = " + x.figuresBeforeDecimal(0));
        System.out.println("fbd (123.456) = " + x.figuresBeforeDecimal(1));
        System.out.println("fbd (123.456) = " + x.figuresBeforeDecimal(2));
        System.out.println("fbd (123.456) = " + x.figuresBeforeDecimal(3));

        x.set(9.76);
        System.out.println("sexp (9.76) = " + x.scaleExp(State.FORMAT_ENG));
        System.out.println("sexp (9.76) = " + x.scaleExp(State.FORMAT_FLOAT));
        x.set(69.76);
        System.out.println("sexp (69.76) = " + x.scaleExp(State.FORMAT_ENG));
        System.out.println("sexp (69.76) = " + x.scaleExp(State.FORMAT_FLOAT));
        x.set(169.76);
        System.out.println("sexp (169.76) = " + x.scaleExp(State.FORMAT_ENG));
        System.out.println("sexp (169.76) = " + x.scaleExp(State.FORMAT_FLOAT));
        x.set(8169.76);
        System.out.println("sexp (8169.76) = " + x.scaleExp(State.FORMAT_ENG));
        System.out.println("sexp (8169.76) = " + x.scaleExp(State.FORMAT_FLOAT));

        // test 6 - cut to given decimal

        x.set(123.456789);
        x.nDigits(1);
        System.out.println("nDigits (1) = " + x.get());
        x.set("123.456789");
        x.nDigits(3);
        System.out.println("nDigits (3) = " + x.get());

        // test 7 - reminder

        x.set(123.456789);
        x.rem(23);
        System.out.println("123.456789 rem 23 = " + x.get());

        for (int k=0; k<8; k++)
            System.out.println("f" + k + ": " + x.formatString(java.util.Locale.US, k));

        x.set(-6.987654);
        for (int k=0; k<8; k++)
            System.out.println("f" + k + ": " + x.formatString(java.util.Locale.US, k));

        x.set(1.1);
        System.out.println(">" + x.formatString(java.util.Locale.US, 8));
        x.set(12.12);
        System.out.println(">" + x.formatString(java.util.Locale.US, 8));
        x.set(123.123);
        System.out.println(">" + x.formatString(java.util.Locale.US, 8));
        x.set(1234.1234);
        System.out.println(">" + x.formatString(java.util.Locale.US, 8));
        x.set(12345.12345);
        System.out.println(">" + x.formatString(java.util.Locale.US, 8));
        x.set(123456.123456);
        System.out.println(">" + x.formatString(java.util.Locale.US, 8));
    }
}
