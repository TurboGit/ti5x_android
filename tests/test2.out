test - 1 0.88598
test - 2 0.08556
orig=10.0
copy=100.0
cos(90)=0.0
cos(45)=0.7071067811867
cos(pi)=-1.0
sqrt(8)=2.828427124746 false
sqrt(0)=0.0 false
sqrt(-5)=2.2360679775 true
log(98765)=4.994603068071 false
ln(98765)=11.50049856996 false
pow(2.5)=316.2277660168 false
one = 1.0
fbd (123.456) = 3
fbd (123.456) = 2
fbd (123.456) = 1
fbd (123.456) = 1
sexp (9.76) = 0
sexp (9.76) = 0
sexp (69.76) = 0
sexp (69.76) = 1
sexp (169.76) = 0
sexp (169.76) = 2
sexp (8169.76) = 3
sexp (8169.76) = 3
nDigits (1) = 123.4
nDigits (3) = 123.456
123.456789 rem 23 = 8.456789
f0: 8
f1: 8.5
f2: 8.46
f3: 8.457
f4: 8.4568
f5: 8.45679
f6: 8.456789
f7: 8.4567890
f0: -7
f1: -7.0
f2: -6.99
f3: -6.988
f4: -6.9877
f5: -6.98765
f6: -6.987654
f7: -6.9876540
>1.10000000
>12.12000000
>123.12300000
>1234.12340000
>12345.12345000
>123456.12345600
